### PR TITLE
I1170 combined gaussian loglikelihood

### PR DESCRIPTION
**Pull request by DavAug**
See PR #1163 

This is probably not the right way to do it, but...

I checked out a previous version of the branch in #1163 and added some changes. When trying do merge this branch into the branch in #1168 I ran into some annoying merge conflicts, so I thought it might be easier to just open a new PR.

### CODE REVIEW

File: pints/_log_likelihoods.py [hunk 1]
```diff
@@ -237,6 +237,228 @@ def __call__(self, x):
         )
 
 
+class CombinedGaussianLogLikelihood(pints.ProblemLogLikelihood):
+    r"""
+    Calculates the log-likelihood for either a
```
Comment (MichaelClerx) [inline pints/_log_likelihoods.py:242]: Would try shortening this a bit and leaving the `:math:` to the second paragraph.
E.g. "Log-likelihood that assumes a mixture of Gaussian noise and Gaussian heteroscedastic noise.". And then the 2nd paragraph gives math, explanation of heteroscedastic, single and multi, etc.

File: pints/_log_likelihoods.py [hunk 2]
```diff
@@ -237,6 +237,228 @@ def __call__(self, x):
         )
 
 
+class CombinedGaussianLogLikelihood(pints.ProblemLogLikelihood):
```
Comment (MichaelClerx) [inline pints/_log_likelihoods.py:240]: "Combined gaussian" sounds much more broad than what this actually is. For example it doesn't combine two N(mu, sigma) log likelihoods

File: pints/_log_likelihoods.py (hunk 1 above)
Comment (MichaelClerx) [reply pints/_log_likelihoods.py:242]: For example, does this generalise to some kind of polynomial normal + normal * model + normal * model^2 + ... etc.?
Needs more specific name!

Comment (DavAug) [reply pints/_log_likelihoods.py:242]: Done

**Comment by MichaelClerx** [comment]
I quite like what it does to the logistic example!

**Comment by DavAug** [comment]
> Needs a more specific name. Other than that I don't really know what this is about so happy for others to review :D

Thank @MichaelClerx !

**Comment by MichaelClerx** [comment]
Is it maybe a linear heteroscedastic gaussian log likelihood?

**Comment by MichaelClerx** [comment]
> Is it maybe a linear heteroscedastic gaussian log likelihood?

There's a certain "a + bx"-ishness about its heteroscedasticity :D

**Comment by DavAug** [comment]
> Is it maybe a linear heteroscedastic gaussian log likelihood?

I'm fine with any name really. I thought because the heteroscedastic noise is called MultiplicativeGaussian, it would be most consistent to call it ConstantAndMultiplicativeGaussian?

**Comment by DavAug** [comment]
> > Is it maybe a linear heteroscedastic gaussian log likelihood?
> 
> There's a certain "a + bx"-ishness about its heteroscedasticity :D

Agreed! Let's see whether other people have string feelings about this. Otherwise I'll go with LinearHeteroscedasticGaussianLogLikelihood

**Comment by MichaelClerx** [comment]
ConstantAndMultiplicative is fine by me too! You decide 😄

File: docs/source/log_likelihoods.rst [hunk 3]
```diff
@@ -19,6 +19,7 @@ Overview:
 - :class:`AR1LogLikelihood`
 - :class:`ARMA11LogLikelihood`
 - :class:`CauchyLogLikelihood`
+- :class:`LinearHeteroscedasticGaussianLogLikelihood`
```
Comment (ben18785) [inline docs/source/log_likelihoods.rst:22]: Sorry, can we make this alphabetical?

File: docs/source/log_likelihoods.rst [hunk 4]
```diff
@@ -35,6 +36,8 @@ Overview:
 
 .. autoclass:: CauchyLogLikelihood
 
+.. autoclass:: LinearHeteroscedasticGaussianLogLikelihood
```
Comment (ben18785) [inline docs/source/log_likelihoods.rst:39]: Same.

File: examples/README.md [hunk 5]
```diff
@@ -82,6 +82,7 @@ relevant code.
 ## Statistical modelling
 - [Autoregressive moving average errors](./stats/autoregressive-moving-average-errors.ipynb)
 - [Cauchy sampling error](./stats/cauchy-sampling-error.ipynb)
+- [Linear heteroscedastic Gaussian error](./stats/linear_heteroscedastic_gaussian_error.ipynb)
```
Comment (ben18785) [inline examples/README.md:85]: Alphabetical too.

File: pints/__init__.py [hunk 6]
```diff
@@ -113,6 +113,7 @@ def version(formatted=False):
     AR1LogLikelihood,
     ARMA11LogLikelihood,
     CauchyLogLikelihood,
+    LinearHeteroscedasticGaussianLogLikelihood,
```
Comment (ben18785) [inline pints/__init__.py:116]: Alphabetical please

File: pints/_log_likelihoods.py [hunk 7]
```diff
@@ -237,6 +237,226 @@ def __call__(self, x):
         )
 
 
+class LinearHeteroscedasticGaussianLogLikelihood(
```
Comment (ben18785) [inline pints/_log_likelihoods.py:240]: Sorry, being dense -- why is it linear? Doesn't eta mean it can be non-linear?

**Comment by DavAug** [comment]
> Looks great -- apart from the few annoying comments about language (sorry!), think we're good to go. For the notebook, because it's hard to comment inline here are some comments here:
> 
> * Should be "logistic growth" not "logistic Growth" in the description
> * The sentence "where the model parameters..." doesn't make sense at the moment
> * I'm not sure we need both the adaptive covariance inference and HMC: I'd just use one of these across your examples. At the moment, the notebook is quite long and, because of that, a bit hard to read.
> * I'd turn off printing to screen in your MCMC running as you have so many of these examples
> * You often put comments using "#" within the blocks rather than write them outside them using Markdown (for example, you have somewhere "Compute diagnostics and show diagnostic plots." within a cell). I'd write them outside as that makes it easier to read.

Ok, cool thanks @ben18785! I'll make the changes.

File: pints/_log_likelihoods.py (hunk 7 above)
Comment (DavAug) [reply pints/_log_likelihoods.py:240]: That is correct! Hmm I don't know why I thought it's linear. It somehow sounded need :D. Anyway it's not correct. Well nonlinear is probably to general. Let's just go back to ConstantAndMultiplicativeGaussianError then?

**Comment by ben18785** [comment]
Yep!

File: docs/source/log_likelihoods.rst (hunk 3 above)
Comment (DavAug) [reply docs/source/log_likelihoods.rst]: done

File: docs/source/log_likelihoods.rst (hunk 4 above)
Comment (DavAug) [reply docs/source/log_likelihoods.rst]: done

File: pints/__init__.py (hunk 6 above)
Comment (DavAug) [reply pints/__init__.py]: done

File: pints/_log_likelihoods.py (hunk 7 above)
Comment (DavAug) [reply pints/_log_likelihoods.py]: done

File: examples/README.md (hunk 5 above)
Comment (DavAug) [reply examples/README.md:85]: done

**Comment by DavAug** [comment]
@MichaelClerx @ben18785 
I've incorporated the suggested changes. I also added a very brief discussion in the notebook how one may deduce that the MCMC run is not successful at first.

Let me know what you think. :)

**Review by ben18785 — CHANGES_REQUESTED** [review]
Looks good @DavAug -- thanks for working through these changes. A couple more things then good to go:

- Can you change the notebook title to "Inference using a two compartment error model incorporating constant and multiplicative errors". Only the first letter "I" should be capitalised.
- There are quite a few comments that start with #s in the cells themselves. Some of these are repetitive: for example, "Infer model parameters with Haario ACMCMC." repeats the text above it. Also, many of these comments at the top of code blocks should be moved outside of them: that'll make them look neater. Can you clean these up?

**Comment by DavAug** [comment]
> Looks good @DavAug -- thanks for working through these changes. A couple more things then good to go:
> 
> * Can you change the notebook title to "Inference using a two compartment error model incorporating constant and multiplicative errors". Only the first letter "I" should be capitalised.
> * There are quite a few comments that start with #s in the cells themselves. Some of these are repetitive: for example, "Infer model parameters with Haario ACMCMC." repeats the text above it. Also, many of these comments at the top of code blocks should be moved outside of them: that'll make them look neater. Can you clean these up?

Sure! Is this how those models are called "two compartment error model"?

**Comment by ben18785** [comment]
@DavAug no (I don't know how those models are called) but perhaps that's a better title?

**Comment by DavAug** [comment]
@ben18785 
Thanks again for the review. I tried to incorporate your change requests. Would you prefer if I removed all comments on the top of the cells?

**Comment by ben18785** [comment]
No need to remove them -- just make the text markdown (i.e. move it above them) as, when they're in the cells, the comments look a bit messy.

**Comment by DavAug** [comment]
> No need to remove them -- just make the text markdown (i.e. move it above them) as, when they're in the cells, the comments look a bit messy.

Sorry I think I'm a little slow right now.. Would you prefer to see
[ ] the requirements for each cell in markdown above the cell, 
[ ] generally when there is a stream of multiple cells a markdown above each cell that contains the short comment that is currently at the top of each cell, or
[ ] both of the above?

**Comment by ben18785** [comment]
Ha, sorry, this is quite hard to communicate and I'm not doing a good job of this. It's easier to explain with a few examples. In general, my point is, you aren't currently making sufficient use of markdown text. The comments written at the top of most cells (but within them) would look much nicer if they were in markdown.

Take the cell immediately below, "Show quantitative and visual diagnostics of MCMC runs". Within the cell, you have a comment, ``This cell needs the HaarioACMC chains'', which is written using #s as it's a comment in Python. This would look better if this comment were moved above the cell (and perhaps explained a little) and written in markdown (immediately below "Show quantitative and visual diagnostics of MCMC runs").

Similarly, take the cell with "Construct posterior" in it. This cell has no markdown comments above it currently. The stuff that's written from "Construct posterior. This cell needs the above defined true parameters, the generated data, and the model wrapper to be able to fix eta" could be moved up to markdown text. It could also perhaps be explained a bit better.

A separate thing that I've just noticed is that you seem to be reloading packages: for example, you import pints a number of times. Can we just have these once at the top (or inline if that makes it easier to understand)?

**Comment by DavAug** [comment]
> Ha, sorry, this is quite hard to communicate and I'm not doing a good job of this. It's easier to explain with a few examples. In general, my point is, you aren't currently making sufficient use of markdown text. The comments written at the top of most cells (but within them) would look much nicer if they were in markdown.
> 
> Take the cell immediately below, "Show quantitative and visual diagnostics of MCMC runs". Within the cell, you have a comment, ``This cell needs the HaarioACMC chains'', which is written using #s as it's a comment in Python. This would look better if this comment were moved above the cell (and perhaps explained a little) and written in markdown (immediately below "Show quantitative and visual diagnostics of MCMC runs").
> 
> Similarly, take the cell with "Construct posterior" in it. This cell has no markdown comments above it currently. The stuff that's written from "Construct posterior. This cell needs the above defined true parameters, the generated data, and the model wrapper to be able to fix eta" could be moved up to markdown text. It could also perhaps be explained a bit better.
> 
> A separate thing that I've just noticed is that you seem to be reloading packages: for example, you import pints a number of times. Can we just have these once at the top (or inline if that makes it easier to understand)?

Ah ok! I'm happy to do those changes :)

The reason why I include the requirements, i.e. "this cell needs...", and the imports in each cell is that I would like to avoid the problem of notebooks that the order of execution matters. Most notebooks only work from top to bottom. This way I highlight which cells need to be executed first, because they compute some variables that are needed. By included the relevant imports in each cell, I avoid that dependence between cells. (that is also why I'd prefer to have the "comment" dependencies ("this cell needs...") within the cell and not as markdown.

I hope that makes sense. But if you have strong feelings about this, I am happy to change it.

**Comment by ben18785** [comment]
Makes sense but I think we, in general, try to have the notebooks so that they run from top to bottom and only import packages once (I'm not sure we always manage to do this though!). So can we change to make it like this?

**Comment by DavAug** [comment]
> Makes sense but I think we, in general, try to have the notebooks so that they run from top to bottom and only import packages once (I'm not sure we always manage to do this though!). So can we change to make it like this?

Yep :)

**Review by ben18785 — APPROVED** [review]
Looks great! Thanks @DavAug

**Comment by ben18785** [comment]
@MichaelClerx Not sure if there were any changes outstanding with this? I'm happy for this to be merged.

**Review by MichaelClerx — APPROVED** [review]

**Comment by MichaelClerx** [comment]
I've merged it in. Thanks all!

**Comment by ben18785** [comment]
Great, thanks Michael!

**Comment by DavAug** [comment]
Thank you Ben and Michael.


I've merged it in. Thanks all!

**Pull request merged**